File: plugins/OpenInEmby/open_in_emby.js
(() => {
  "use strict";

  // ═══════════════════════════════════════════════════════════
  // 常量定义
  // ═══════════════════════════════════════════════════════════

  const PLUGIN_ID = "open_in_emby";
  const ICON_URL = "https://cdn.jsdelivr.net/gh/lige47/QuanX-icon-rule@main/icon/04ProxySoft/emby.png";

  const TYPE_PATTERNS = {
    scene: /\/scenes\/(\d+)/,
    performer: /\/performers\/(\d+)/,
    studio: /\/studios\/(\d+)/
  };

  const EMBY_TYPE_MAP = {
    scene: "Movie",
    performer: "Person",
    studio: "BoxSet"
  };

  // ═══════════════════════════════════════════════════════════
  // 模块 1: 按钮管理（UI 层）
  // ═══════════════════════════════════════════════════════════

  let cachedIcon = null;

  function getIconNode() {
    if (cachedIcon) return cachedIcon.cloneNode(true);

    const img = document.createElement("img");
    img.src = ICON_URL;
    img.style.cssText = "display:block;width:1.2em;height:1.2em;";

    img.onload = () => { cachedIcon = img.cloneNode(true); };

    img.onerror = () => {
      const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
      svg.setAttribute("viewBox", "0 0 512 512");
      svg.style.cssText = "display:block;width:1.2em;height:1.2em;";
      svg.innerHTML = `<circle cx="256" cy="256" r="200" fill="#52B54B"/><text x="256" y="320" font-size="200" text-anchor="middle" fill="white" font-weight="bold">E</text>`;
      cachedIcon = svg;
      img.replaceWith(svg.cloneNode(true));
    };

    return img;
  }

  function initCSS() {
    if (document.getElementById("open-in-emby-css")) return;

    const style = document.createElement("style");
    style.id = "open-in-emby-css";
    style.textContent = `
      .open-in-emby-btn {
        background: none !important;
        border: none !important;
        padding: 0 !important;
        cursor: pointer;
        display: inline-flex;
        align-items: center;
        justify-content: center;
      }
      .open-in-emby-btn:hover {
        background: none !important;
      }
      .open-in-emby-btn img,
      .open-in-emby-btn svg {
        display: block;
        width: 1.2em;
        height: 1.2em;
        transition: opacity 0.2s;
      }
      .open-in-emby-btn:hover img,
      .open-in-emby-btn:hover svg {
        opacity: 0.7;
      }
    `;
    document.head.appendChild(style);
  }

  function createEmbyButton(type, id) {
    const btn = document.createElement("button");
    btn.className = "open-in-emby-btn";
    btn.appendChild(getIconNode());

    btn.onclick = async () => {
      try {
        const settings = await getPluginSettings();
        if (!settings.embyServer || !settings.embyInternalServer || !settings.embyApiKey) {
          throw new Error("请先配置 Emby 服务器地址、内网地址和 API Key");
        }

        const data = await gql(`mutation RunPluginOperation($plugin_id: ID!, $args: Map) { runPluginOperation(plugin_id: $plugin_id, args: $args) }`, {
          plugin_id: PLUGIN_ID,
          args: {
            embyServer: settings.embyServer,
            embyInternalServer: settings.embyInternalServer,
            embyApiKey: settings.embyApiKey,
            stash_id: id.toString(),
            includeItemTypes: EMBY_TYPE_MAP[type] || 'Movie',
          },
        });

        const result = (typeof data?.runPluginOperation === 'string' ? JSON.parse(data.runPluginOperation) : data?.runPluginOperation) || {};

        if (result?.success && result?.url) {
          window.open(result.url, "_blank", "noopener,noreferrer");
        } else {
          throw new Error(result?.error || "Emby 中未找到匹配");
        }
      } catch (err) {
        alert(`Open in Emby: ${err.message}`);
      }
    };

    return btn;
  }

  function getItemId() {
    const path = window.location.pathname;
    const hash = window.location.hash || "";
    for (const [type, regex] of Object.entries(TYPE_PATTERNS)) {
      let m = path.match(regex) || hash.match(regex);
      if (m) return { type, id: parseInt(m[1], 10) };
    }
    return null;
  }

  function getCardInfo(card) {
    const link = card.querySelector('a[href*="/scenes/"], a[href*="/performers/"], a[href*="/studios/"]');
    if (!link) return null;
    for (const [type, regex] of Object.entries(TYPE_PATTERNS)) {
      const m = link.href.match(regex);
      if (m) return { type, id: parseInt(m[1], 10) };
    }
    return null;
  }

  function addSceneDetailButton() {
    const item = getItemId();
    if (!item || item.type !== 'scene') return;

    const toolbarGroups = Array.from(document.querySelectorAll("span.scene-toolbar-group"));
    const toolbarGroup = toolbarGroups.length > 1 ? toolbarGroups[1] : null;
    if (!toolbarGroup || document.getElementById("open_in_emby__btn")) return;

    const btn = createEmbyButton('scene', item.id);
    btn.id = "open_in_emby__btn";

    const viewCountSpan = toolbarGroup.querySelector('span:has(.view-count-button), span:has(.count-button)');
    if (viewCountSpan && viewCountSpan.parentNode) {
      viewCountSpan.parentNode.insertBefore(btn, viewCountSpan);
    } else {
      toolbarGroup.appendChild(btn);
    }
  }

  function addPerformerStudioButton() {
    const item = getItemId();
    if (!item || (item.type !== 'performer' && item.type !== 'studio')) return;
    if (document.getElementById("open-in-emby-toolbar")) return;

    const qualityGroup = document.querySelector('.quality-group');
    if (!qualityGroup) return;

    const btn = createEmbyButton(item.type, item.id);
    btn.id = "open-in-emby-toolbar";

    const oCounter = qualityGroup.querySelector('.o-counter-button, .count-button');
    if (oCounter && oCounter.parentNode) {
      oCounter.parentNode.insertBefore(btn, oCounter.nextSibling);
    } else {
      qualityGroup.appendChild(btn);
    }
  }

  function addCardButton(card) {
    if (!card || card.querySelector('.open-in-emby-btn')) return;

    const info = getCardInfo(card);
    if (!info) return;

    let refBtn;
    if (info.type === 'scene') {
      refBtn = card.querySelector('.organized button, .organized-btn');
    } else if (info.type === 'performer') {
      refBtn = card.querySelector('.scene-count button');
    } else if (info.type === 'studio') {
      refBtn = card.querySelector('.performer-count button');
    }
    if (!refBtn) return;

    const parent = refBtn.closest('.btn-group, .card-popovers');
    if (parent) {
      const btn = createEmbyButton(info.type, info.id);
      parent.appendChild(btn);
    }
  }

  // ═══════════════════════════════════════════════════════════
  // 模块 2: GraphQL 通信和配置读取
  // ═══════════════════════════════════════════════════════════

  async function gql(query, variables) {
    const res = await fetch(localStorage.getItem("apiEndpoint") || "/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": localStorage.getItem("apiKey") ? `Bearer ${localStorage.getItem("apiKey")}` : "",
      },
      credentials: "include",
      body: JSON.stringify({ query, variables }),
    });
    const json = await res.json();
    if (json?.errors?.length) throw new Error(json.errors[0].message);
    return json.data;
  }

  async function getPluginSettings() {
    const data = await gql(`query { configuration { plugins } }`);
    const cfg = data?.configuration?.plugins?.[PLUGIN_ID] || {};
    const get = (k) => {
      const v = cfg[k];
      return (v && typeof v === 'object' && 'value' in v) ? v.value : v;
    };
    return {
      embyServer: (get('embyServer') || "").trim().replace(/\/+$/, ""),
      embyInternalServer: (get('embyInternalServer') || "").trim().replace(/\/+$/, ""),
      embyApiKey: (get('embyApiKey') || "").trim(),
    };
  }

  // ═══════════════════════════════════════════════════════════
  // 模块 3: 初始化和监听
  // ═══════════════════════════════════════════════════════════

  function init() {
    initCSS();

    // 首次执行
    addSceneDetailButton();
    addPerformerStudioButton();
    document.querySelectorAll('.scene-card, .performer-card, .studio-card').forEach(addCardButton);

    // 路由监听（页面切换）
    if (window.PluginApi?.Event?.addEventListener) {
      PluginApi.Event.addEventListener("stash:location", () => {
        setTimeout(() => {
          addSceneDetailButton();
          addPerformerStudioButton();
        }, 200);
      });
    }

    // DOM 监听（刷新/动态加载）
    const observer = new MutationObserver(() => {
      addSceneDetailButton();
      addPerformerStudioButton();

      document.querySelectorAll('.scene-card, .performer-card, .studio-card')
        .forEach(card => {
          if (!card.querySelector('.open-in-emby-btn')) {
            addCardButton(card);
          }
        });
    });
    observer.observe(document.body, { childList: true, subtree: true });
  }

  init();
})();

File: plugins/OpenInEmby/open_in_emby.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
open_in_emby.py - Open in Emby 后端代理

通过 Stash 后端代理 Emby API 请求，绕过浏览器 CSP 限制。
前端读取配置并传递给后端。
"""

import json
import sys
import urllib.request
import urllib.parse
import urllib.error


def query_emby(emby_server: str, emby_internal_server: str, emby_api_key: str, stash_id: str, include_item_types: str = "Movie") -> dict:
    """通过 Stash 本地 ID 匹配 Emby 视频"""
    formatted_id = f"stash.{stash_id}"
    params = urllib.parse.urlencode({
        "IncludeItemTypes": include_item_types,
        "AnyProviderIdEquals": formatted_id,
        "Recursive": "True",
        "api_key": emby_api_key,
    })

    # ⭐ 使用内网地址搜索
    emby_url = f"{emby_internal_server}/emby/Items?{params}"

    try:
        req = urllib.request.Request(emby_url)
        req.add_header("Accept", "application/json")

        with urllib.request.urlopen(req, timeout=30) as response:
            data = json.loads(response.read().decode("utf-8"))
            items = data.get("Items", [])

            if items:
                item = items[0]
                item_id = item.get("Id")
                server_id = item.get("ServerId", "")
                # ⭐ 使用外网地址构建跳转 URL
                if server_id:
                    detail_url = f"{emby_server}/web/index.html#!/item?id={item_id}&serverId={server_id}"
                else:
                    detail_url = f"{emby_server}/web/index.html#!/item?id={item_id}"
                return {
                    "success": True,
                    "error": None,
                    "url": detail_url,
                    "item": {"id": item_id, "name": item.get("Name")},
                }
            else:
                return {"success": True, "error": "未找到匹配", "url": None, "item": None}

    except urllib.error.HTTPError as e:
        return {"success": False, "error": f"HTTP 错误 {e.code}", "url": None, "item": None}
    except urllib.error.URLError as e:
        return {"success": False, "error": f"网络错误：{e.reason}", "url": None, "item": None}
    except Exception as e:
        return {"success": False, "error": f"未知错误：{str(e)}", "url": None, "item": None}


def main():
    """主函数"""
    # 读取 stdin 的 JSON 输入
    try:
        input_json = sys.stdin.read()

        if not input_json:
            print(json.dumps({"output": {"success": False, "error": "输入为空"}}))
            return

        input_data = json.loads(input_json)
    except json.JSONDecodeError as e:
        print(json.dumps({"output": {"success": False, "error": f"JSON 解析失败：{e}"}}))
        return
    except Exception as e:
        print(json.dumps({"output": {"success": False, "error": f"读取失败：{e}"}}))
        return

    # 从 args 获取所有参数（前端传递）
    args = input_data.get("args", {})

    emby_server = args.get("embyServer", "")
    emby_internal_server = args.get("embyInternalServer", "")
    emby_api_key = args.get("embyApiKey", "")
    stash_id = args.get("stash_id")
    include_item_types = args.get("includeItemTypes", "Movie")

    # 验证参数
    if not stash_id:
        print(json.dumps({"output": {"success": False, "error": "缺少参数：stash_id"}}))
        return

    if not emby_server or not emby_internal_server or not emby_api_key:
        print(json.dumps({"output": {"success": False, "error": "缺少配置：Emby 服务器地址、内网地址、API Key"}}))
        return

    # 查询 Emby
    result = query_emby(emby_server, emby_internal_server, emby_api_key, str(stash_id), include_item_types)

    # 输出结果
    print(json.dumps({"output": result}))


if __name__ == "__main__":
    main()
